File: sdks/rust/src/providers/minimax/http.rs
use super::{
    protocol::{self, failure, Error, Packet},
    url, SynthesisItem,
};
use crate::{
    generated::minimax_output::*,
    http::{HttpRequest, HttpResponse, HttpTransport, TransportError},
    runtime::StreamingInput,
    sse::Decoder,
};
use std::{
    future::{poll_fn, Future},
    pin::Pin,
    task::{Context, Poll},
};

type SubtitleFuture<'a> =
    Pin<Box<dyn Future<Output = Result<Vec<MiniMaxTimestamp>, TransportError>> + Send + 'a>>;
pub(super) struct HttpStream<'a> {
    transport: &'a dyn HttpTransport,
    body: Option<StreamingInput<Vec<u8>>>,
    status: u16,
    retry_after: Option<String>,
    decoder: Option<Decoder>,
    bytes: std::vec::IntoIter<u8>,
    document: Vec<u8>,
    limit: usize,
    timing: Option<&'static str>,
    final_packet: Option<Packet>,
    subtitle: Option<SubtitleFuture<'a>>,
    received: bool,
    subtitles_done: bool,
    pub done: bool,
}
impl<'a> HttpStream<'a> {
    pub fn new(
        response: HttpResponse,
        transport: &'a dyn HttpTransport,
        timing: Option<&'static str>,
        limit: usize,
    ) -> Result<Self, TransportError> {
        let sse = (200..300).contains(&response.status)
            && response.headers.iter().any(|(k, v)| {
                k.eq_ignore_ascii_case("content-type")
                    && v.split(';')
                        .next()
                        .unwrap_or("")
                        .trim()
                        .eq_ignore_ascii_case("text/event-stream")
            });
        let retry_after = response
            .headers
            .iter()
            .find(|(k, _)| k.eq_ignore_ascii_case("retry-after"))
            .map(|(_, v)| v.clone());
        Ok(Self {
            transport,
            body: Some(response.body),
            status: response.status,
            retry_after,
            decoder: if sse {
                Some(Decoder::new(limit)?)
            } else {
                None
            },
            bytes: vec![].into_iter(),
            document: vec![],
            limit,
            timing,
            final_packet: None,
            subtitle: None,
            received: false,
            subtitles_done: false,
            done: false,
        })
    }
    fn error(&self, p: Packet) -> TransportError {
        Box::new(Error {
            message: if p.message.is_empty() {
                format!("MiniMax HTTP {}", self.status)
            } else {
                p.message
            },
            code: p.code,
            status: Some(self.status),
            retry_after: self.retry_after.clone(),
        })
    }
    fn accept(&mut self, mut p: Packet) -> Result<Option<SynthesisItem>, TransportError> {
        if p.code.is_some() {
            return Err(self.error(p));
        }
        if self.decoder.is_some() {
            if p.status.is_none() {
                return Err(failure("MiniMax SSE audio is missing data.status"));
            }
        } else if p.status != Some(2) {
            return Err(failure("MiniMax JSON synthesis did not report completion"));
        }
        let audio = p.audio.take().filter(|v| !v.is_empty());
        let trace = p.trace.clone();
        if p.status == Some(2) {
            self.final_packet = Some(p);
        }
        Ok(audio.map(|audio| {
            self.received = true;
            if self.timing.is_some() {
                let mut e = protocol::envelope(true, trace);
                e.audio = Some(audio);
                SynthesisItem::OrderedOrTimeline(e)
            } else {
                SynthesisItem::Bytes(audio)
            }
        }))
    }
    pub fn next(
        &mut self,
        cx: &mut Context<'_>,
    ) -> Poll<Result<Option<SynthesisItem>, TransportError>> {
        for _ in 0..4096 {
            if self.final_packet.is_some() {
                break;
            }
            if let Some(byte) = self.bytes.next() {
                if let Some(event) = self.decoder.as_mut().unwrap().push(byte)? {
                    if event.data == "[DONE]" {
                        return Poll::Ready(Err(failure(
                            "MiniMax HTTP stream ended before completion",
                        )));
                    }
                    let p = protocol::packet(&event.data)?;
                    if let Some(item) = self.accept(p)? {
                        return Poll::Ready(Ok(Some(item)));
                    }
                }
                continue;
            }
            match self.body.as_mut().unwrap().as_mut().poll_next(cx) {
                Poll::Pending => return Poll::Pending,
                Poll::Ready(Some(Err(e))) => return Poll::Ready(Err(e)),
                Poll::Ready(Some(Ok(bytes))) => {
                    if self.decoder.is_some() {
                        self.bytes = bytes.into_iter();
                    } else {
                        if bytes.len() > self.limit - self.document.len() {
                            return Poll::Ready(Err(failure(
                                "MiniMax response exceeds max_json_bytes",
                            )));
                        }
                        self.document.extend(bytes);
                    }
                }
                Poll::Ready(None) => {
                    self.body = None;
                    if self.decoder.is_some() {
                        return Poll::Ready(Err(failure(
                            "MiniMax HTTP stream ended before completion",
                        )));
                    }
                    let text = protocol::text(&self.document)?;
                    if !(200..300).contains(&self.status) {
                        let p = match crate::json::Raw::parse(text) {
                            Ok(_) => protocol::packet(text)?,
                            Err(_) => Packet::default(),
                        };
                        return Poll::Ready(Err(self.error(p)));
                    }
                    let p = protocol::packet(text)?;
                    self.document = Vec::new();
                    if let Some(item) = self.accept(p)? {
                        return Poll::Ready(Ok(Some(item)));
                    }
                }
            }
        }
        let Some(final_packet) = &self.final_packet else {
            cx.waker().wake_by_ref();
            return Poll::Pending;
        };
        self.body = None;
        self.bytes = vec![].into_iter();
        if let Some(decoder) = &mut self.decoder {
            decoder.finish();
        }
        if !self.received {
            return Poll::Ready(Err(failure("MiniMax returned no audio")));
        }
        if let Some(kind) = self.timing.filter(|_| !self.subtitles_done) {
            if self.subtitle.is_none() {
                let target = final_packet
                    .subtitle
                    .as_deref()
                    .filter(|v| !v.is_empty())
                    .ok_or_else(|| failure("MiniMax omitted requested subtitles"))?;
                self.subtitle = Some(Box::pin(download(
                    self.transport,
                    url(target, "", false)?,
                    self.limit,
                    kind,
                )));
            }
            match self.subtitle.as_mut().unwrap().as_mut().poll(cx) {
                Poll::Pending => return Poll::Pending,
                Poll::Ready(result) => {
                    self.subtitle = None;
                    let mut e = protocol::envelope(true, final_packet.trace.clone());
                    e.timestamps = result?;
                    self.subtitles_done = true;
                    return Poll::Ready(Ok(Some(SynthesisItem::OrderedOrTimeline(e))));
                }
            }
        }
        let p = self.final_packet.take().unwrap();
        self.done = true;
        Poll::Ready(Ok(Some(SynthesisItem::Done(MiniMaxDoneEvent {
            event: Default::default(),
            trace_id: p.trace,
            usage: p.usage,
        }))))
    }
}

async fn download(
    transport: &dyn HttpTransport,
    target: String,
    limit: usize,
    kind: &str,
) -> Result<Vec<MiniMaxTimestamp>, TransportError> {
    // Independent files must not inherit API credentials or transport cookies.
    let mut response = transport
        .send(HttpRequest {
            method: "GET".into(),
            url: target,
            headers: vec![],
            body: vec![],
        })
        .await?;
    if !(200..300).contains(&response.status) {
        return Err(Box::new(Error {
            message: "MiniMax subtitle download failed".into(),
            status: Some(response.status),
            code: None,
            retry_after: None,
        }));
    }
    let mut data = Vec::new();
    while let Some(chunk) = poll_fn(|cx| response.body.as_mut().poll_next(cx)).await {
        let chunk = chunk?;
        if chunk.len() > limit - data.len() {
            return Err(failure("MiniMax response exceeds max_json_bytes"));
        }
        data.extend(chunk);
        let mut yielded = false;
        poll_fn(|cx| {
            if yielded {
                Poll::Ready(())
            } else {
                yielded = true;
                cx.waker().wake_by_ref();
                Poll::Pending
            }
        })
        .await;
    }
    protocol::subtitles(protocol::text(&data)?, kind)
}

File: sdks/go/providers/minimax/http.go
package minimax

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	out "github.com/speechswitch/client/sdks/go/generated/minimax_output"
	"github.com/speechswitch/client/sdks/go/runtime"
	"io"
	"net/http"
	"strings"
	"sync"
)

type httpStream struct {
	parent, ctx                                      context.Context
	cancel                                           context.CancelFunc
	stop                                             func() bool
	transport                                        runtime.HTTPTransport
	response                                         *runtime.HTTPResponse
	body                                             runtime.Input[[]byte]
	limit                                            int
	timing                                           string
	once                                             sync.Once
	mutex, lifecycle                                 sync.Mutex
	closed                                           bool
	closeError                                       error
	initialized, received, subtitles, done, terminal bool
	decoder                                          *runtime.SSEDecoder
	buffer                                           []byte
	final                                            *packet
}

func (s *httpStream) closeResources() error {
	s.once.Do(func() {
		s.cancel()
		s.lifecycle.Lock()
		s.closed = true
		body := s.body
		s.body = nil
		s.lifecycle.Unlock()
		if body != nil {
			s.closeError = body.Close()
		}
	})
	return s.closeError
}
func (s *httpStream) Close() error { s.stop(); return s.closeResources() }
func (s *httpStream) replaceBody(body runtime.Input[[]byte]) error {
	s.lifecycle.Lock()
	old, closed := s.body, s.closed
	if !closed {
		s.body = body
	}
	s.lifecycle.Unlock()
	if old != nil {
		_ = old.Close()
	}
	if closed && body != nil {
		_ = body.Close()
		return s.ctx.Err()
	}
	return nil
}
func readDocument(ctx context.Context, body runtime.Input[[]byte], limit int) ([]byte, error) {
	var data []byte
	for {
		chunk, err := body.Next(ctx)
		if err == io.EOF {
			return bytes.TrimPrefix(data, []byte{239, 187, 191}), nil
		}
		if err != nil {
			return nil, err
		}
		if len(chunk) > limit-len(data) {
			return nil, errors.New("MiniMax response exceeds MaxJSONBytes")
		}
		data = append(data, chunk...)
	}
}
func (s *httpStream) failure(p packet) *Error {
	code := runtime.Optional[int]{}
	if p.code != 0 {
		code = runtime.Some(p.code)
	}
	retry := runtime.Optional[string]{}
	if values, ok := s.response.Header["Retry-After"]; ok && len(values) > 0 {
		retry = runtime.Some(values[0])
	}
	message := p.message
	if message == "" {
		message = fmt.Sprintf("MiniMax HTTP %d", s.response.StatusCode)
	}
	return &Error{Message: message, Code: code, StatusCode: s.response.StatusCode, RetryAfter: retry}
}
func (s *httpStream) Next(ctx context.Context) (out.SynthesisItem, error) {
	stop := context.AfterFunc(ctx, func() { s.closeResources() })
	defer stop()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if s.terminal || s.done {
		return nil, io.EOF
	}
	item, err := s.next(ctx)
	if cause := ctx.Err(); cause != nil {
		item, err = nil, cause
	} else if cause := s.parent.Err(); cause != nil {
		item, err = nil, cause
	} else if s.ctx.Err() != nil && !s.done {
		item, err = nil, io.EOF
	}
	if err != nil {
		s.terminal = true
		s.Close()
	}
	return item, err
}
func (s *httpStream) audio(p packet) out.SynthesisItem {
	s.received = true
	if s.timing == "" {
		return out.SynthesisItemAsBytes{Value: p.audio}
	}
	return out.SynthesisItemAsOrderedOrTimeline{Value: out.MiniMaxEnvelope{Correlation: out.MiniMaxEnvelopeCorrelationAsTimeline{}, Audio: runtime.Some(p.audio), Timestamps: []out.MiniMaxTimestamp{}, TraceId: p.trace}}
}
func (s *httpStream) next(ctx context.Context) (out.SynthesisItem, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if s.ctx.Err() != nil {
		return nil, io.EOF
	}
	if !s.initialized {
		s.initialized = true
		isSSE := strings.ToLower(strings.TrimSpace(strings.SplitN(s.response.Header.Get("Content-Type"), ";", 2)[0])) == "text/event-stream"
		if s.response.StatusCode < 200 || s.response.StatusCode >= 300 || !isSSE {
			data, err := readDocument(ctx, s.response.Body, s.limit)
			if err != nil {
				return nil, err
			}
			if s.response.StatusCode < 200 || s.response.StatusCode >= 300 {
				if !json.Valid(data) {
					return nil, s.failure(packet{})
				}
				p, err := decodePacket(data)
				if err != nil {
					return nil, err
				}
				return nil, s.failure(p)
			}
			p, err := decodePacket(data)
			if err != nil {
				return nil, err
			}
			if p.code != 0 {
				return nil, s.failure(p)
			}
			if p.status != 2 {
				return nil, errors.New("MiniMax JSON synthesis did not report completion")
			}
			s.final = &p
			if len(p.audio) > 0 {
				return s.audio(p), nil
			}
		} else {
			var err error
			s.decoder, err = runtime.NewSSEDecoder(s.limit)
			if err != nil {
				return nil, err
			}
		}
	}
	for s.final == nil {
		if len(s.buffer) == 0 {
			data, err := s.response.Body.Next(ctx)
			if err == io.EOF {
				return nil, errors.New("MiniMax HTTP stream ended before completion")
			}
			if err != nil {
				return nil, err
			}
			s.buffer = data
		}
		for len(s.buffer) > 0 {
			event, err := s.decoder.Push(s.buffer[0])
			s.buffer = s.buffer[1:]
			if err != nil {
				return nil, err
			}
			if event == nil {
				continue
			}
			if event.Data == "[DONE]" {
				return nil, errors.New("MiniMax HTTP stream ended before completion")
			}
			p, err := decodePacket([]byte(event.Data))
			if err != nil {
				return nil, err
			}
			if p.code != 0 {
				return nil, s.failure(p)
			}
			if p.status != 1 && p.status != 2 {
				return nil, errors.New("MiniMax SSE audio is missing data.status")
			}
			if p.status == 2 {
				s.final = &p
			}
			if len(p.audio) > 0 {
				return s.audio(p), nil
			}
			if s.final != nil {
				break
			}
		}
	}
	if s.decoder != nil {
		s.decoder.Finish()
	}
	s.buffer = nil
	if err := s.replaceBody(nil); err != nil {
		return nil, err
	}
	if !s.received {
		return nil, errors.New("MiniMax returned no audio")
	}
	if s.timing != "" && !s.subtitles {
		if !s.final.subtitle.Present || s.final.subtitle.Value == "" {
			return nil, errors.New("MiniMax omitted requested subtitles")
		}
		target, err := endpoint(s.final.subtitle.Value, "", false)
		if err != nil {
			return nil, err
		}
		request, err := http.NewRequestWithContext(s.ctx, http.MethodGet, target, nil)
		if err != nil {
			return nil, err
		}
		// Never copy synthesis credentials onto the independent file request.
		response, err := runtime.OpenResponse(request, s.transport)
		if err != nil {
			return nil, err
		}
		if err = s.replaceBody(response.Body); err != nil {
			return nil, err
		}
		if response.StatusCode < 200 || response.StatusCode >= 300 {
			return nil, &Error{Message: "MiniMax subtitle download failed", StatusCode: response.StatusCode}
		}
		data, err := readDocument(ctx, response.Body, s.limit)
		if err != nil {
			return nil, err
		}
		marks, err := decodeSubtitles(data, s.timing)
		if err != nil {
			return nil, err
		}
		if err = s.replaceBody(nil); err != nil {
			return nil, err
		}
		s.subtitles = true
		return out.SynthesisItemAsOrderedOrTimeline{Value: out.MiniMaxEnvelope{Correlation: out.MiniMaxEnvelopeCorrelationAsTimeline{}, Timestamps: marks, TraceId: s.final.trace}}, nil
	}
	s.done = true
	s.Close()
	return out.SynthesisItemAsDone{Value: out.MiniMaxDoneEvent{TraceId: s.final.trace, Usage: s.final.usage}}, nil
}
